### PR TITLE
[Bugfix] fix recompile in qwen3 vl

### PR DESCRIPTION
**Pull request by narutolhy**
## Motivation

Qwen3-VL automatically injects input_deepstack_embeds into the language model inputs when requests contain mm_inputs (multimodal inputs). The current implementation does not handle this injected tensor consistently, so the model execution path differs between multimodal and text-only requests. In mixed traffic, this leads to unnecessary TorchDynamo recompilations at runtime (e.g., repeatedly specializing on the presence/absence of deepstack inputs), increasing latency and reducing throughput stability.

This PR makes the deepstack injection path explicit and stable to avoid redundant recompilation for Qwen3-VL multimodal inference.

## Modifications

Standardized the Qwen3-VL deepstack injection behavior so that the language model forward path remains stable when mm_inputs is present.

Ensured input_deepstack_embeds is handled in a consistent way across requests to prevent runtime recompilation churn (especially under mixed multimodal/text workloads).

Refactored the relevant code paths to make the deepstack injection explicit and easier to reason about.

## Accuracy Tests

The update in test/manual/nightly/test_vlms_vit_cuda_graph.py case covers the accuracy.

## Checklist

- [X] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [X] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - `/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`
4. After green CI and required approvals, ask Merge Oncalls to merge.

### CODE REVIEW

File: python/sglang/srt/model_executor/piecewise_cuda_graph_runner.py [hunk 1]
```diff
@@ -357,11 +369,18 @@ def warmup_torch_compile(self, num_tokens: int):
...
-                forward_batch.input_ids,
-                forward_batch.positions,
-                forward_batch,
-            )
+            kwargs = {}
+            if self.is_multimodal and hasattr(self, "input_deepstack_embeds"):
+                kwargs["input_deepstack_embeds"] = self.input_deepstack_embeds[
+                    :num_tokens
+                ]
+            with torch.no_grad():
```
Comment (Oasis-Git) [inline python/sglang/srt/model_executor/piecewise_cuda_graph_runner.py]: Which part of vlm will introduce non no_grad() mode?

File: python/sglang/srt/model_executor/piecewise_cuda_graph_runner.py [hunk 2]
```diff
@@ -683,12 +707,19 @@ def replay(
                 self.moe_layers,
             ):
                 with set_compiled(True):
-                    output = self.model_runner.model.forward(
-                        static_forward_batch.input_ids,
-                        static_forward_batch.positions,
-                        static_forward_batch,
-                        **kwargs,
-                    )
+                    call_kwargs = {}
```
Comment (Oasis-Git) [inline python/sglang/srt/model_executor/piecewise_cuda_graph_runner.py]: Can we move all the kwargs preparation outside the piecewise cuda graph context?

File: python/sglang/srt/models/qwen3_vl.py [hunk 3]
```diff
@@ -566,19 +566,41 @@ def __init__(
...
+                )
+                self.deepstack_embeds_buffer = torch.zeros(
+                    new_len, total, dtype=dtype, device=device
+                ).contiguous()
+            sep = self.hidden_size * layer_idx
+            return self.deepstack_embeds_buffer[:seq_len, sep : sep + self.hidden_size]
         sep = self.hidden_size * layer_idx
         return input_deepstack_embeds[:, sep : sep + self.hidden_size]
 
+    @torch.no_grad()
```
Comment (Oasis-Git) [inline python/sglang/srt/models/qwen3_vl.py]: With this no_grad() maybe we can remove the one in pcg file??

**Review by Oasis-Git — COMMENTED** [review]
Previously when I run this model I noticed the recompilation is caused by grad mode. So could you please share which part introduce forward with gradient?

**Comment by yuan-luo** [comment]
> > Could you help to verify with SGLANG_VIT_ENABLE_CUDA_GRAPH=1? Qwen3-VL ViT CUDA Graph considered injecting deepstack output into LLM layers, just want make sure the full ViT+LLM CUDA Graph path working correctly.
> > The issue also exists when SGLANG_VIT_ENABLE_CUDA_GRAPH is set to 0.
> 
> Hi @yuan-luo Thank you for your review. I tested with VIT CUDA Graph before. It has some error. This is the issue #16883 I think this is due to the latest optimizations, but after I removed them, it resulted in a CUDA address access error during graph replay. I think it's because different graphs are being used for the same `seq_len` length, and your planned refactoring should solve this problem. #16902 I'll try test it with your new pr. Thank you

@narutolhy Did https://github.com/sgl-project/sglang/pull/16902 fix your problem?

File: python/sglang/srt/models/qwen3_vl.py (hunk 3 above)
Comment (narutolhy) [reply python/sglang/srt/models/qwen3_vl.py]: Yes, I removed it.

File: python/sglang/srt/model_executor/piecewise_cuda_graph_runner.py (hunk 2 above)
Comment (narutolhy) [reply python/sglang/srt/model_executor/piecewise_cuda_graph_runner.py]: OK, I removed them outside the piecewise cuda graph context

File: python/sglang/srt/model_executor/piecewise_cuda_graph_runner.py (hunk 1 above)
Comment (narutolhy) [reply python/sglang/srt/model_executor/piecewise_cuda_graph_runner.py]: Removed

**Review by Oasis-Git — APPROVED** [review]
LGTM from the pcg side. @yuan-luo Could you take a next step look please? Thanks
